docs/design-patterns.md

```diff
@@ -0,0 +1,152 @@
+# Design Patterns
+
+The main operating principle for `vdom` is:
+
+> Write functions that return `vdom` elements
+
+In matching with `React` parlance, we'll call these functions components. This allows you to share, remix, and use components amongst everyone. You'll be able to compose together components to create greater layouts than you have
+before.
+
+## Introductory component
+
+We'll start with a component that takes a level of happiness that produces a light visualization:
+
+```python
+from vdom.helpers import div, span, p, meter
+
+def happiness(level=90):
+  smiley = "😃"
+  percent = level / 100.
+    
+  if(percent < 0.25):
+    smiley = "☹️"
+  elif(percent < 0.50):
+    smiley = "😐"
+  elif(percent < 0.75):
+    smiley = "😀"
+
+    
+  return span(
+      p('Happiness ', smiley),
+      meter(level, min=0, low=25, high=75, optimum=90, max=100, value=level)
+  )
+```
+
+The user of the component only has to know how to call `happiness` with their level of happiness:
+
+```python
+happiness(96)
+```
+
+<span>
+<p>Happiness 😃</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="96">
+  96
+</meter>
+</span>
+
+------------
+
+:tada: Our first component is ready! Since we can think of these as little building block components, we can put several of these together to create whole layouts:
+
+```python
+div(
+  happiness(10),
+  happiness(32),
+  happiness(65),
+  happiness(80)
+)
+```
+
+<span>
+<p>Happiness ☹️</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="10">
+  10
+</meter>
+</span>
+<span>
+<p>Happiness 😐</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="32">
+  32
+</meter>
+</span>
+<span>
+<p>Happiness 😀</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="65">
+  65
+</meter>
+</span>
+<span>
+<p>Happiness 😃</p>
+<meter min="0" low="25" high="75" optimum="90" max="100" value="80">
+  80
+</meter>
+</span>
+
+-------------------
+
+
+## Working with python objects
+
+For this section, you'll need `ggplot` and `matplotlib`. We'll create a component
```

Comment:
`matplotlib` packages installed. We'll create a component, `fancy_hist`, that ... side: